### ISSUE TITLE
Allow node modules to have node modules as dependencies

### ISSUE THREAD

**Issue by knolleary** [node-red/node-red#569]
Node-red scans for `node_modules` directories starting from its install location and working up the tree to the root path. When it finds such a directory, it examines the `package.json` files of the modules in the directories immediately below; it doesn't scan deeper for additional `node_module` directories.

This means a node module cannot declare another node module as a dependency and have it load properly - which is a reasonable thing to allow.

- when a valid node-red node module is discovered, scan its `node_modules` directory
- ensure we are recording the version number of node modules
- if a node module is found that we have already loaded:
  - if the version number matches, skip it and carry on.
  - if the version numbers don't match, refuse to load it (and if this was in a node module's node_module directory, refuse to load the top level module as its dependency cannot be satisfied)

**Comment by TotallyInformation** [comment]
+1

Not sure about that last bullet though. Might there be edge cases where the package.json might indicate that the existing loaded module might actually satisfy even though the direct load might clash? For example, if the existing module was loaded some time ago but not updated and the new module tried to load the latest version but could actually use the older one? Or maybe I'm just over complicating things (again!).

**Comment by coolchip** [comment]
This would be very helpful. 
At the moment I develop a node, that uses the serial connection. Everything I need is in node-red-node-serialport. I still can use this node (configutation and functionality). But it has to be installed for node-red. If I simply put it in my package.json, node-red don't find it.
Is there any way to declare node-red-node-dependencies. 
Perhaps in the package.json. Something like this:

```json
"dependencies": {
    "nodered-node-serialport": "^0.0.1"
},
"keywords": [ "node-red" ],
"node-red"     : {
    "nodes": {
        "sample": "sample/sample.js"
    },
    "dependencies": {
        "nodered-node-serialport": "^0.0.1"
    }
}
```

**Comment by knolleary** [comment]
@coolchip as this is still an open issue, nothing has changed from the original problem statement - there is no way today for a node to declare other node-red nodes as dependencies.

**Comment by zaphod-42** [comment]
This topic is 2 years old and I still can't find anything, in my case we are working on a suite of modules that will share configuration nodes (similar to what coolchip was saying). I'm sure this isn't the best way to do this, but if anyone has any thoughts I'd like to hear them:

`function ensureDependencies(deps){
	//deps is just an array of dependencies that should be NPM installable, if no version is
	//required, and array of string may be sent, if a version is needed, an array of objects:
	//{name: "module-name", version: "^3.0.0"}
	var registry = require(require.main.filename.replace('red.js', 'red/runtime/nodes/registry/index.js'));
	var modules = registry.getModuleList();
	deps.forEach((mod) => {
		var o = typeof mod == "object";
		var m = o ? mod.name : mod;
		var v = o ? mod.verison : false;
		if(typeof modules[m] == 'undefined'){
			registry.installModule(m, v).then().catch((err) => {
				console.log("ensureDependencies Failed to install "+m+" from "+module.filename);
			});
		}
	});
}`

**Comment by zaphod-42** [comment]
To be clear, I use it in the module file outside of the constructor:

```
module.exports = function(RED){
  ensureDependencies(['serial-port']);
  function SerialDevice(config){
    RED.nodes.createNode(this, config);
  }
  RED.nodes.registerType("serial-device", SerialDevice); 
}
```

**Comment by hardillb** [comment]
If you are building a suite of nodes just include them all in the same npm module. The node-red section of the package.json allows you to declare multiple Node-RED nodes

**Comment by zaphod-42** [comment]
My company manufactures iot devices, we have around 900 on our site, the plan is to group them into categories that make sense so you can grab 5-10 nodes to control a variety of devices (like a bunch of temperature/humidity sensors) packing them all into one module seems like overkill

**Comment by knolleary** [comment]
A big difference today is that npm doesn't bury node module dependencies; it tries to lift modules to the highest possible directory so it can be shared.

I think that means you should just be able to declare the dependencies as you would nornally and it will simply work.

Certainly worth trying to see where we stand.

**Comment by zaphod-42** [comment]
No such luck, I just tried it and the nodes aren't installed via dependencies in the package.json.

**Comment by zaphod-42** [comment]
Although in this case I am just using npm link for testing, once I'm ready to publish one of these I'll test it out

**Comment by dceejay** [comment]
you can also use `npm pack` to package your code locally into an installable tgz file without having to publish it. You can then install it with `npm i your_package_name.tgz`

**Comment by zaphod-42** [comment]
wow thanks for that lol, I hadn't ever seen pack. I just tried it also with a dependency, and while the dependency was installed, the nodes are not available in node red

**Comment by knolleary** [comment]
One problem with what I said previously - whilst npm may layout the nodes so node-red can discover them, that will only work if node-red is restarted to trigger a scan. If installed via the Palette Manager in the editor, the runtime will only know to try to load the module being installed - not any of its dependencies.

**Comment by zaphod-42** [comment]
I tried that, node-red doesn't seem to discover any nodes that are included as part of a dependency, regardless of restarts.

**Comment by knolleary** [comment]
It will depend entirely on where npm put the files. What version of npm are you using?

When this issue was first raised, if you installed module A and module C that both depended on B, npm 1.x or 2.x would give you:

```
.node-red
    \- node_modules
           |- A
           |    \- node_modules
           |           \- B
           \- C
                \- node_modules
                       \- B
```

And node-red would only scan one level of node_modules directories so you'd get A and C, not B.

With npm 3.x and later, you will get something like:

```
.node-red
    \- node_modules
           |- A
           |- B
           \- C
```

And Node-RED will be able to discover them all.

That's the theory - until you commented on this issue, this wasn't a subject I'd put any thought into for some time. So would be interested to hear how npm is laying out your modules when you install them.

**Comment by zaphod-42** [comment]
I'm currently using npm 1.4.21 because that is what the version of node on raspbian stretch used out of the box. I'll test it on another setup in a bit with npm 3.x, if that works this function should never have to do anything because the modules would already be installed, but I will probably still package my modules with this for people who don't update everything to the latest versions.

**Comment by knolleary** [comment]
Raspbian Stretch doesn't include npm by default - a fact we are well aware of as it relates to the packaging of node-red on Raspbian. The version of node in Raspbian is about to drop out of all support - not even getting security fixes.

This is why we provide our [update script](https://nodered.org/docs/hardware/raspberrypi) to get users onto the LTS node version, with corresponding npm.

**Comment by zaphod-42** [comment]
sorry, I should have been more clear. Because stretch uses an antiquated node version, 1.4.21 is the newest npm that can be installed without first upgrading node

**Comment by dceejay** [comment]
see here for support info - https://github.com/nodejs/Release
I'd recommend upgrading as per our docs.

**Comment by zaphod-42** [comment]
I've seen that, and completely agree, my Pi at home is upgraded. My only concern is backwards compatibility, if it was just for me none of this would matter, this is just something to handle dependencies in something that is distributed to other people. If I put a giant link at the top of the readme with red characters saying this upgrade is required, I will still have to point it out to people (I've been programming for over a decade and am intimately familiar with customers tendency to ignore warning labels).

Aside from the fact that the new versions of npm handle this in a way that should alleviate the issue, is there a particular problem with using this method as a fallback for dependency management?

**Comment by knolleary** [comment]
If you're referring to the code you shared above...
```
  var registry = require(require.main.filename.replace('red.js', 'red/runtime/nodes/registry/index.js'));
  var modules = registry.getModuleList();
```
That is delving into the internals of Node-RED and uses non-public apis. I cannot guarantee things won't move around as we work to stabilise things on our way to version 1.0.

**Comment by zaphod-42** [comment]
Fair enough, I knew that was a possibility, I guess I will add a warning label and possible a notification if npm isn't up to date. Thanks for the time, and the work... most of my last few years have been spent bouncing between C++ and PHP, when we found node-red it completely refocused half of our development, Javascript has been my favorite language since... well the late 90's ;) And this flow editor is one of the coolest visual implementations of IoT and coding that I've ever seen.

**Comment by dceejay** [comment]
So OK to close this as it's trying to workaround a problem caused by unsupported version of npm ?

**Comment by knolleary** [comment]
No. There's still this issue to address: https://github.com/node-red/node-red/issues/569#comment-373515258

**Comment by tve** [comment]
+1: I'm running into the same issue where one module needs a config node from another module and while the latter is installed at the npm level its nodes are not scanned/discovered.

**Comment by tve** [comment]
Just to provide a bit of perspective. Due to this bug, to install FlexDash using the palette manager I would have to tell users to manually install node-red-fd-corewidgets and node-red-flexdash and node-red-flexdash-plugin. Most likely, there's also going to be a node-red-flexdash-sidebar to add to the list. If a user installs the corewidgets but forgets to install one of the other packages then at a first glance everything seems to work, except that config nodes are missing and therefore the input fields in the flow editor don't work. The net result are nodes that seemingly are there but somehow don't work.

Instead, I will have to recommend to install using npm, where just installing node-red-fd-corewidgets before launching node-red brings in everything else and works properly.

This is particularly a shame when using docker, because instead of just something like
```
docker run --rm -ti -p 1880:1880 nodered/node-red:2.2.2
```
and then using the flow-manager to get FlexDash one has to run:
```
docker run --rm -ti -p 1880:1880 --entrypoint bash nodered/node-red:2.2.2 \
  -c "npm i @flexdash/node-red-fd-corewidgets; npm start --cache /data/.npm -- -v -userDir /data"
```

**Comment by dirkjanfaber** [comment]
I too am developing a node that has a dependency. In my case it relies on the availability of the config node from `node-red-contrib-modbus`.
I guess for me it would be enough to have the ability to force a rescan to be called from the `post-install` part in the `package.json`.

**Comment by knolleary** [comment]
I've been thinking about how best to enable support for this.

My original proposal was based on the behaviour of `npm` at the time to always install dependencies beneath the module being installed. These days, it tries to flatten modules so they could appear in adjacent directories. That makes it less convenient to find what has been installed.

We also have to consider transitive dependencies - `A` depends on `B`, which depends on `C`. We need to load `C` then `B` then `A`.

Triggering a full rescan is a lot of work for the runtime to do. I think we can be a bit smarter - with some amount of cooperation of nodes that want to have 

What if, under the `node-red` section of `package.json` we have an extra `dependencies` section that lists *only* the node-red modules it depends on. (Not all of the other non-node-red modules that appears in the regular deps section).

Then, when we load that module, we will know exactly what node-red modules it depends on and can ensure they are present and loaded.

For example:

```
    "dependencies": {
        "node-red-node-random": "1.2.3"
    },
    "node-red"      : {
        "nodes"     : {
            "rbe": "rbe.js"
        },
        "dependencies": [
            "node-red-node-random"
        ]
    },
```

The downside is a bit of duplication between the normal `dependencies` section. But it makes it much more deterministic.

**Comment by dirkjanfaber** [comment]
This looks like a workable and nice solution to me.

**Comment by Steve-Mcl** [comment]
Nick, would these dependencies follow the same semantic versioning as regular deps?

What happens if node-a has a node-red.dependancy of node-b AND node-b has a node-red.dependancy of node-a?

**Comment by Steve-Mcl** [comment]
@tve have you had a chance to catch up on the suggestion made by Nick above?

Would appreciate your feedback please.

**Comment by TotallyInformation** [comment]
Just caught up on this. It is something I'd like to make use of. I agree with Steve's question about using standard npm dependency properties though. Indeed, I suspect that might even make it easier to deliver.

Don't forget that the dependency value can be one of quite a number of different things that encompasses source location, branch/version, just version, ...

**Comment by tve** [comment]
Makes sense to me and I'll echo the caveat that it would be better if the new dependencies section had the same format as the normal one, i.e., allowed versions to be specified.

**Comment by knolleary** [comment]
I proposed the array to avoid duplication of information - and remove the possibility where there could be a mismatch in what version is specified in the two locations.

Regardless of that, we do have the issue to consider where two modules ask for different versions of a particular dependency. Unlike normal npm modules, Node-RED cannot load two different versions at the same time. We will load whatever is found first on the module path.

**Comment by tve** [comment]
Oh, I didn't understand that the version was taken from the normal dependencies. In that case, this ought to work for me at least.

**Comment by Steve-Mcl** [comment]
> Oh, I didn't understand that the version was taken from the normal dependencies. In that case, this ought to work for me at least.

Same here. Yes, this looks like a good way to satisfy requirement.
